### PR TITLE
added hashmap

### PR DESCRIPTION
**Pull request by ArtemkaDev**

### CODE REVIEW

File: src/blockchaintree.rs [hunk 1]
```diff
@@ -788,10 +792,10 @@ impl BlockChainTree {
         Ok(())
     }
 
-    pub fn get_derivative_chain(
+    pub async fn get_derivative_chain(
         &mut self,
         addr: &[u8; 33],
-    ) -> Result<Option<Box<DerivativeChain>>, BlockChainTreeError> {
+    ) -> Result<Option<&Arc<RwLock<DerivativeChain>>>, BlockChainTreeError> {
```
Comment (YeahNotSewerSide) [inline src/blockchaintree.rs]: Should be `Result<Option<Arc<RwLock<DerivativeChain>>>, BlockChainTreeError>`
Arc should be cloned for response

File: src/blockchaintree.rs [hunk 2]
```diff
@@ -800,12 +804,13 @@ impl BlockChainTree {
...
             let result = DerivativeChain::new(&path_string).change_context(
-                BlockChainTreeError::BlockChainTree(BCTreeErrorKind::GetDerivChain),
-            )?;
-
-            let chain = Box::new(result);
+                    BlockChainTreeError::BlockChainTree(BCTreeErrorKind::GetDerivChain),
+                )?;
+            
+            
+            self.deratives.insert(*addr, Arc::new(RwLock::new(result)));
```
Comment (YeahNotSewerSide) [inline src/blockchaintree.rs]: No checks if derivative chain already exists in the hashmap

File: src/blockchaintree.rs [hunk 3]
```diff
@@ -815,11 +820,12 @@ impl BlockChainTree {
...
     }
 
-    pub fn create_derivative_chain(
+    pub async fn create_derivative_chain(
+        &mut self,
         addr: &[u8; 33],
         genesis_hash: &[u8; 32],
         global_height: u64,
-    ) -> Result<Box<DerivativeChain>, BlockChainTreeError> {
+    ) -> Result<&Arc<RwLock<DerivativeChain>>, BlockChainTreeError> {
```
Comment (YeahNotSewerSide) [inline src/blockchaintree.rs]: `Result<Option<Arc<RwLock<DerivativeChain>>>, BlockChainTreeError>`

File: src/blockchaintree.rs [hunk 4]
```diff
@@ -852,14 +858,17 @@ impl BlockChainTree {
...
-
+        
         chain
             .dump_config(&root_path)
             .change_context(BlockChainTreeError::BlockChainTree(
                 BCTreeErrorKind::CreateDerivChain,
             ))?;
 
-        Ok(Box::new(chain))
+        self.deratives.insert(*addr, Arc::new(RwLock::new(chain)));
```
Comment (YeahNotSewerSide) [inline src/blockchaintree.rs]: No checks if derivative chain already exists in the hashmap

File: src/blockchaintree.rs [hunk 5]
```diff
@@ -852,14 +858,17 @@ impl BlockChainTree {
...
+        
+
+        Ok(self.deratives.get(addr).unwrap())
```
Comment (YeahNotSewerSide) [inline src/blockchaintree.rs]: Better to use [entry](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry)

File: src/blockchaintree.rs (hunk 1 above)
Comment (ArtemkaDev) [reply src/blockchaintree.rs]: i can't do that
image

Comment (ArtemkaDev) [reply src/blockchaintree.rs]: maybe  `Result<Option<Arc<RwLock<DerivativeChain>>>`  thats beter

Comment (YeahNotSewerSide) [reply src/blockchaintree.rs]: > Should be `Result<Option<Arc<RwLock<DerivativeChain>>>`, BlockChainTreeError> Arc should be cloned for response

I fixed markdown in this line

File: src/blockchaintree.rs (hunk 3 above)
Comment (ArtemkaDev) [reply src/blockchaintree.rs:828]: image
thats also need <>

Comment (ArtemkaDev) [reply src/blockchaintree.rs]: image
Need <> and arguments

Comment (YeahNotSewerSide) [reply src/blockchaintree.rs]: > `Result<Option<Arc<RwLock<DerivativeChain>>>, BlockChainTreeError>`

fixed markdown

**Pull request merged**